lxd/storage/block/discard.go

```diff
@@ -0,0 +1,264 @@
+package block
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"strconv"
+
+	"github.com/canonical/lxd/shared"
+)
+
+// ClearBlock fully resets a block device or disk file using the most efficient mechanism available.
+// For files, it will truncate them down to zero and back to their original size.
+// For blocks, it will attempt a variety of discard options, validating the result with marker files and eventually
+// fallback to full zero-ing.
+//
+// An offset can be specified to only reset a part of a device.
+func ClearBlock(blockPath string, blockOffset int64) error {
+	// Open the block device for checking.
+	fd, err := os.OpenFile(blockPath, os.O_RDWR, 0644)
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			// If the file is missing, there is nothing to clear.
+			return nil
+		}
+
+		return err
+	}
+
+	defer fd.Close()
+
+	// Get the size of the file/block.
+	size, err := fd.Seek(0, io.SeekEnd)
```

Comment:
Should it error out if the retrieved `size` is identical to the provided `blockOffset`?

Reply:
im not sure. Will the function run OK, if so i dont see an issue with this? Could you expand on your concern here? Thanks

Reply:
If both sizes are identical, it means no clearing will actually happen so the caller likely fed the wrong offset. Same as what I flagged with the fat LVM case where the LV was already resized when block clearing was requested for the new size (instead of the old size).

Reply:
right i see, ill update to error on that scenario thanks

Reply:
have updated